### PR TITLE
Implement Tenderly Web3 Actions to monitor different RedStone smart contracts

### PR DESCRIPTION
**Pull request by matjazv**
### What was the problem?

This PR resolves #LISK-1157.

### How was it solved?

- Web3 Actions for `USDT/USD`, `LSK/USD` and `ETH/USD` token pairs were created to monitor their update
- It's possible to deploy those actions using Tenderly CLI

### How was it tested?

Web3 Actions for all three token pairs were deployed to Tenderly and they were manually executed to check if alerts inside Opsgenie were created.

### CODE REVIEW

File: web3-actions/priceFeedMonitor/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# RedStone Price Feed Monitoring
+
+Inside this folder, you will find the code for the RedStone Price Feed monitoring using Tenderly Web3 Actions. This code is used to monitor the price feed of the RedStone smart contracts for different token pairs and trigger alerts if the price feed was not updated for a certain period of time. In this case, the alert is sent to the Opsgenie.
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]:
```suggestion
Inside this directory, you will find the code for the RedStone Price Feed monitoring using Tenderly Web3 Actions. This code is used to monitor the price feed of the RedStone smart contracts for different token pairs and trigger alerts if the price feed was not updated for a certain period of time. In this case, the alert is sent to Opsgenie.
```

File: web3-actions/priceFeedMonitor/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Install Tenderly CLI
+If you haven't already, install [Tenderly CLI](https://github.com/Tenderly/tenderly-cli#installation).
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]:
```suggestion
If you haven't already, install [Tenderly CLI](https://github.com/Tenderly/tenderly-cli#installation).
```

File: web3-actions/priceFeedMonitor/README.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+tenderly login
+```
+
+## Build and Publish/Deploy Web3 Actions
+
+Before you can build and publish/deploy the Web3 Actions, you need modify the configuration `.yaml` file for the project.
+Some configuration files inside this folder are:
+- `lskUsd.yaml` - configuration file for the RedStone price feed monitoring for the LSK/USD token pair
+- `ethUSD.yaml` - configuration file for the RedStone price feed monitoring for the ETH/USD token pair
+- `usdtUsd.yaml` - configuration file for the RedStone price feed monitoring for the USDT/USD token pair
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]:
```suggestion
- [`lskUsd.yaml`](./lskUsd.yaml) - configuration file for the RedStone price feed monitoring for the LSK/USD token pair
- [`ethUsd.yaml`](./ethUsd.yaml) - configuration file for the RedStone price feed monitoring for the ETH/USD token pair
- [`usdtUsd.yaml`](./usdtUsd.yaml) - configuration file for the RedStone price feed monitoring for the USDT/USD token pair
```

File: web3-actions/priceFeedMonitor/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+You need to provide the following information in the configuration file:
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]:
```suggestion
You need to provide the following information in the configuration file(s), under `actions`:
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/package.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "devDependencies": {
+        "typescript": "^4.3.5"
+    },
+    "dependencies": {
+        "@tenderly/actions": "^0.2.0",
+        "@types/node": "^20.8.10",
+        "opsgenie-sdk": "^0.5.1"
+    },
+    "private": true
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/package.json]:
```suggestion
    }
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                name: "Lisk"
+            }
+        ],
+        visibleTo: [
+            {
+                name: "Lisk",
+                type: "team"
+            }
+        ],
+        priority: "P1"
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts]:
```suggestion
        priority: "P2"
```

File: web3-actions/priceFeedMonitor/README.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+- `YOUR_USERNAME` - your Tenderly account
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]: Also, please update the config yaml files to reflect the same 🙏 
```suggestion
- `YOUR_ACCOUNT_SLUG` - your Tenderly account
```

File: web3-actions/priceFeedMonitor/README.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+- `YOUR_PROJECT_SLUG` - your Tenderly project
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md:24]:
```suggestion
- `YOUR_PROJECT_SLUG` - your Tenderly project

**Note**: Both of the above values can be accessed from the Setting page on your Tenderly dashboard.
```

File: web3-actions/priceFeedMonitor/README.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+```bash
+tenderly actions build --project-config [yaml_file]
+```
+
+To publish/deploy the Web3 Actions, run the following command:
+
+```bash
+tenderly actions publish --project-config [yaml_file]
+or
+tenderly actions deploy --project-config [yaml_file]
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]:
```suggestion
tenderly actions build --project-config [yaml_filename_without_extension]
\```

To publish/deploy the Web3 Actions, run the following command:

\```bash
tenderly actions publish --project-config [yaml_filename_without_extension]
or
tenderly actions deploy --project-config [yaml_filename_without_extension]
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/package.json [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    "name": "tokenPairMonitor",
+    "scripts": {
+        "build": "tsc"
+    },
+    "devDependencies": {
+        "typescript": "^4.3.5"
+    },
+    "dependencies": {
+        "@tenderly/actions": "^0.2.0",
+        "@types/node": "^20.8.10",
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/package.json:11]: Please re-generate the package-lock.json after adding the following dependency 🙏 
```suggestion
        "@types/node": "^20.8.10",
        "ethers": "^5.7.2",
```

File: web3-actions/priceFeedMonitor/ethUsd.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+    specs:
+      RedStonePriceFeed_ETH-USD:
+        description: Check if token pair price for ETH/USD is updated at least each 6 hours.
+        function: ethUsd:monitorEthUsdFn
+        trigger:
+          type: periodic
+          periodic:
+            interval: 1h
+        execution_type: parallel
+project_slug: ""
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/ethUsd.yaml]: `account_id` and `project_slug` could also be removed from all the yaml files.

```suggestion
actions:
  YOUR_USERNAME/YOUR_PROJECT_SLUG:
    runtime: v2
    sources: tokenPairMonitor
    specs:
      RedStonePriceFeed_ETH-USD:
        description: Check if token pair price for ETH/USD is updated at least each 6 hours.
        function: ethUsd:monitorEthUsdFn
        trigger:
          type: periodic
          periodic:
            interval: 1h
        execution_type: parallel
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    };
+
+    opsgenie.configure({
+        host: 'https://api.opsgenie.com',
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts]: Define the URL as a constant above.

File: web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        const latestRoundData = await contract.latestRoundData();
+
+        // Log the results
+        console.log("Latest Round Data for", tokenPair, "token pair:");
+        console.log("Answer:", latestRoundData.answer.toString());
+        console.log("Started At:", latestRoundData.startedAt.toString());
+        console.log("Updated At:", latestRoundData.updatedAt.toString());
+
+        // Check if updatedAt is older than 6 hours from the current time (+2 minutes buffer for Web3 Actions)
+        const sixHoursInSeconds = 6 * 60 * 60 + 120;
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts]: As discussed offline.
```suggestion
        // Check if updatedAt is older than 6 hours from the current time
        const sixHoursInSeconds = 6 * 60 * 60;
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        const updatedAt = Number(latestRoundData.updatedAt.toString());
+
+        console.log("Current Time:", currentTimestamp);
+
+        if (currentTimestamp - updatedAt > sixHoursInSeconds) {
+            console.warn("Warning: The latest data for", tokenPair, "token pair is older than 6 hours.");
+            // Send alert to the Opsgenie
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts]:
```suggestion
        if ((currentTimestamp - updatedAt) > sixHoursInSeconds) {
            console.warn("Warning: The latest data for", tokenPair, "token pair is older than 6 hours.");
            // Send alert to Opsgenie
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/commonCode.ts:1]: The file can be maybe renamed to `validate.ts`?

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

**Comment by matjazv** [comment]
@sameersubudhi 
All your comments were resolved inside commit https://github.com/LiskHQ/lisk-contracts/pull/212/commits/f1ce796d2d93583979d1196e60ace8b262293224.

File: web3-actions/priceFeedMonitor/README.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+```bash
+tenderly login
+```
+
+## Build and Publish/Deploy Web3 Actions
+
+Before you can build and publish/deploy the Web3 Actions, you need modify the configuration `.yaml` file for the project.
+Some configuration files inside this directory are:
+- [`lskUsd.yaml`](./lskUsd.yaml) - configuration file for the RedStone price feed monitoring for the LSK/USD token pair
+- [`ethUSD.yaml`](./ethUsd.yaml) - configuration file for the RedStone price feed monitoring for the ETH/USD token pair
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/README.md]: Incorrect display text.
```suggestion
- [`ethUsd.yaml`](./ethUsd.yaml) - configuration file for the RedStone price feed monitoring for the ETH/USD token pair
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/ethUsd.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+	PeriodicEvent,
+} from '@tenderly/actions';
+import { checkTokenPairPriceUpdateTime } from './validate';
+
+// Define the contract address and token pair
+const CONTRACT_ADDRESS = "0x6b7AB4213c77A671Fc7AEe8eB23C9961fDdaB3b2";
+const TOKEN_PAIR = "ETH/USD";
+
+export const monitorEthUsdFn: ActionFn = async (context: Context, event: Event) => {
+	let periodicEvent = event as PeriodicEvent;
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/ethUsd.ts]:
```suggestion
	const periodicEvent = event as PeriodicEvent;
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/lskUsd.ts [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+	PeriodicEvent,
+} from '@tenderly/actions';
+import { checkTokenPairPriceUpdateTime } from './validate';
+
+// Define the contract address and token pair
+const CONTRACT_ADDRESS = "0xa1EbA9E63ed7BA328fE0778cFD67699F05378a96";
+const TOKEN_PAIR = "LSK/USD";
+
+export const monitorLskUsdFn: ActionFn = async (context: Context, event: Event) => {
+	let periodicEvent = event as PeriodicEvent;
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/lskUsd.ts]:
```suggestion
	const periodicEvent = event as PeriodicEvent;
```

File: web3-actions/priceFeedMonitor/tokenPairMonitor/usdtUsd.ts [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+	PeriodicEvent,
+} from '@tenderly/actions';
+import { checkTokenPairPriceUpdateTime } from './validate';
+
+// Define the contract address and token pair
+const CONTRACT_ADDRESS = "0xd2176Dd57D1e200c0A8ec9e575A129b511DBD3AD";
+const TOKEN_PAIR = "USDT/USD";
+
+export const monitorUsdtUsdFn: ActionFn = async (context: Context, event: Event) => {
+	let periodicEvent = event as PeriodicEvent;
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/tokenPairMonitor/usdtUsd.ts]:
```suggestion
	const periodicEvent = event as PeriodicEvent;
```

File: web3-actions/priceFeedMonitor/ethUsd.yaml [hunk 19]
```diff
@@ -0,0 +1,13 @@
+actions:
+  YOUR_ACCOUNT_SLUG/YOUR_PROJECT_SLUG:
+    runtime: v2
+    sources: tokenPairMonitor
+    specs:
+      RedStonePriceFeed_ETH-USD:
+        description: Check if token pair price for ETH/USD is updated at least each 6 hours.
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/ethUsd.yaml]:
```suggestion
        description: Check if token pair price for ETH/USD is updated at least once every 6 hours.
```

File: web3-actions/priceFeedMonitor/lskUsd.yaml [hunk 20]
```diff
@@ -0,0 +1,13 @@
+actions:
+  YOUR_ACCOUNT_SLUG/YOUR_PROJECT_SLUG:
+    runtime: v2
+    sources: tokenPairMonitor
+    specs:
+      RedStonePriceFeed_LSK-USD:
+        description: Check if token pair price for LSK/USD is updated at least each 6 hours.
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/lskUsd.yaml]:
```suggestion
        description: Check if token pair price for LSK/USD is updated at least once every 6 hours.
```

File: web3-actions/priceFeedMonitor/usdtUsd.yaml [hunk 21]
```diff
@@ -0,0 +1,13 @@
+actions:
+  YOUR_ACCOUNT_SLUG/YOUR_PROJECT_SLUG:
+    runtime: v2
+    sources: tokenPairMonitor
+    specs:
+      RedStonePriceFeed_USDT-USD:
+        description: Check if token pair price for USDT/USD is updated at least each 6 hours.
```
Comment (sameersubudhi) [inline web3-actions/priceFeedMonitor/usdtUsd.yaml]:
```suggestion
        description: Check if token pair price for USDT/USD is updated at least once every 6 hours.
```

**Review by sameersubudhi — APPROVED** [review]
Looks good overall. However, please consider the following minor observations.

**Review by Incede — APPROVED** [review]

**Pull request merged**